### PR TITLE
Pipeoutput

### PR DESCRIPTION
**Pull request by chme**
Hi @ejurgensen,

this is a my first attempt at adding another output type. It allows forked-daapd to ouput data to a named pipe (fifo). A usecase is for example to support the visualizer in ncmcpp.

Currently with this pr there is only support for one pipe output and it outputs the data immediately upon receiving  with no synchronization (this is something that can be added later).

The second commit deletes the legacy alsa output. If you left it in the source tree on purpose, i will drop that commit, just let me know.

One thing i am not sure about the outputs/device implementation is, if setting the device-id to a fixed value is correct (i chose device-id = 1) or if this may lead to conflicting ids.

### CODE REVIEW

**Comment by ejurgensen** [comment]
Looks good to me. Good that you also remove laudio_alsa.c, I had simply forgotten it.

You are right about device_id, since they are set by the output there is a risk of conflict. This risk is very small currently. For laudio, alsa and dummy will use 0, while pulseaudio will use 0, 1, 2, 3... and up. Raop and Chromecast both use large id's, which could in principle conflict, but the likelihood is low, so I haven't done anything about it.

Since device_id 1 is likely to conflict with Pulseaudio, you should use something else. The lazy solution is to just use whatever sufficiently high random number you can think of. You are also welcome to look at more proper solution for achieving uniqueness (a prefix perhaps, but might have implications for the raop or chromecast implementations?).

**Comment by chme** [comment]
I took the easy path and fifo output now has the fixed id of 100. 

I will think about a proper solution, but my first guess is that this will require introducing a new id field in the speakers table and mapping the "internal" id to the raop/cast ids.

File: src/outputs/fifo.c [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+
+  device = calloc(1, sizeof(struct output_device));
+  if (!device)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Out of memory for fifo device\n");
+      return -1;
+    }
+
+  device->id = 100;
+  device->name = nickname;
```
Comment (ejurgensen) [inline src/outputs/fifo.c]: You need to strdup here because it will be freed by outputs_device_free().

File: src/outputs/fifo.c [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+static void
+fifo_empty(struct fifo_session *fifo_session)
+{
+  char buf[FIFO_BUFFER_SIZE];
+  int bytes = 1;
+
+  while (bytes > 0 && errno != EINTR)
+    bytes = read(fifo_session->input_fd, buf, FIFO_BUFFER_SIZE);
```
Comment (ejurgensen) [inline src/outputs/fifo.c]: flush can be called without the fifo having been created, which can lead to a crash. Here is the output from Valgrind:

==16132== Invalid read of size 4
==16132==    at 0x808B65D: read (unistd.h:44)
==16132==    by 0x808B65D: fifo_empty.isra.0 (fifo.c:185)
==16132==    by 0x808B6E9: fifo_flush (fifo.c:386)
==16132==    by 0x80841F6: outputs_flush (outputs.c:194)
==16132==    by 0x807FB9E: playback_stop (player.c:2215)
==16132==    by 0x80973A7: command_cb_sync (commands.c:84)
==16132==    by 0x80973A7: command_cb (commands.c:141)
==16132==    by 0x5913233: event_base_loop (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x5913FE2: event_base_dispatch (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x807F2BD: player (player.c:3938)
==16132==    by 0x5C5EF71: start_thread (pthread_create.c:312)
==16132==    by 0x5D5FF8D: clone (clone.S:129)
==16132==  Address 0x8 is not stack'd, malloc'd or (recently) free'd
==16132== 
==16132== 
==16132== Process terminating with default action of signal 11 (SIGSEGV)
==16132==  Access not within mapped region at address 0x8
==16132==    at 0x808B65D: read (unistd.h:44)
==16132==    by 0x808B65D: fifo_empty.isra.0 (fifo.c:185)
==16132==    by 0x808B6E9: fifo_flush (fifo.c:386)
==16132==    by 0x80841F6: outputs_flush (outputs.c:194)
==16132==    by 0x807FB9E: playback_stop (player.c:2215)
==16132==    by 0x80973A7: command_cb_sync (commands.c:84)
==16132==    by 0x80973A7: command_cb (commands.c:141)
==16132==    by 0x5913233: event_base_loop (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x5913FE2: event_base_dispatch (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x807F2BD: player (player.c:3938)
==16132==    by 0x5C5EF71: start_thread (pthread_create.c:312)
==16132==    by 0x5D5FF8D: clone (clone.S:129)

File: src/outputs/fifo.c [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+static int
+fifo_flush(output_status_cb cb, uint64_t rtptime)
+{
+  struct fifo_session *fifo_session = sessions;
+
+  fifo_empty(fifo_session);
+
+  fifo_session->status_cb = cb;
```
Comment (ejurgensen) [inline src/outputs/fifo.c]: fifo_session might be NULL (see above)

File: src/outputs/fifo.c [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+static int
+fifo_make(struct fifo_session *fifo_session)
+{
+  if (mkfifo(fifo_session->path, 0666) < 0)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Could not create FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return -1;
+    }
+
```
Comment (ejurgensen) [inline src/outputs/fifo.c]: Consider a DBG log message here telling that the FIFO was created (like you have for removing)

**Comment by chme** [comment]
Updated the commit. fifo_empty still should be called with a session != NULL, but i added a check in fifo_flush.

Thanks for the review!

**Comment by ejurgensen** [comment]
Merged, thanks!

**Pull request merged**